File: algorithms/270.Closest_Binary_Search_Tree_Value/Closest_Binary_Search_Tree_Value.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def closestValue(self, root, target):
        """
        :type root: TreeNode
        :type target: float
        :rtype: int
        """
        # basically in-order travesal and find the closest
        prev = 0x7fffffffffffffff
        result = 0x7fffffffffffffff
        path = [root]
        while path:
            cur = path[-1]
            if cur:
                path.append(cur.left)
            else:
                path.pop(-1)
                if path:
                    cur = path.pop(-1)
                    curdist = abs(cur.val - target)
                    if curdist > prev:
                        return result
                    else:
                        result = cur.val
                        prev = curdist
                    path.append(cur.right)
        return result
        

File: algorithms/270.Closest_Binary_Search_Tree_Value/Closest_Binary_Search_Tree_Value.cc
class Solution {
public:
    int closestValue(TreeNode* root, double target) {
        TreeNode* node = root;
        int res = node->val;
        while(node) {
            if(abs(target - node->val) < abs(target - res)) {
                res = node->val;
            }

            if(target < node->val) {
                node = node->left;
            } else {
                node = node->right;
            }
        }
        
        return res;
    }
};
// Closest Binary Search Tree Value
class Solution {
public:
  int closestValue(TreeNode *x, double target) {
    int opt;
    double optd = numeric_limits<double>::max();
    while (x) {
      double d = fabs(target-x->val);
      if (d < optd)
        optd = d, opt = x->val;
      x = target < x->val ? x->left : x->right;
    }
    return opt;
  }
};